File: notifications.js
var notifLang = document.documentElement.lang || "en";
var notifFilterCat = "";
var notifAutoTimer = null;

function notifT(en, ta) { return notifLang === "ta" ? ta : en; }

function notifToast(msg, type) {
    var t = document.getElementById("notif-toast");
    if (!t) return;
    t.textContent = msg; t.className = "notif-toast " + type;
    t.classList.add("show");
    clearTimeout(t._hide);
    t._hide = setTimeout(function () { t.classList.remove("show"); }, 3000);
}

function notifSetLanguage() {
    fetch("/set-language/" + (notifLang === "en" ? "ta" : "en"), { method: "POST" })
        .then(function () { location.reload(); });
}

/* Boot */
document.addEventListener("DOMContentLoaded", function () {
    notifGenerate();
    notifAutoTimer = setInterval(notifLoadList, 120000);
});

/* Generate notifications from user data */
function notifGenerate() {
    fetch("/api/notifications/generate", { method: "POST" })
        .then(function (r) { return r.json(); })
        .then(function (res) {
            notifLoadList();
        })
        .catch(function () {});
}

/* Load notification list */
function notifLoadList() {
    var p = new URLSearchParams();
    if (notifFilterCat) p.set("category", notifFilterCat);
    p.set("all", "1");
    fetch("/api/notifications?" + p.toString())
        .then(function (r) { return r.json(); })
        .then(function (res) {
            if (!res.success) return;
            notifRenderList(res.notifications);
            document.getElementById("notif-total-count").textContent = res.total || 0;
            document.getElementById("notif-unread-count").textContent = res.unread_count || 0;
        })
        .catch(function () {});
}

/* Filter by category */
function notifFilter(cat) {
    notifFilterCat = cat;
    var tabs = document.querySelectorAll(".notif-cat-tab");
    tabs.forEach(function (t) { t.classList.toggle("active", t.dataset.cat === cat); });
    notifLoadList();
}

/* Render notification cards */
function notifRenderList(notifications) {
    var list = document.getElementById("notif-list");
    if (!list) return;
    if (!notifications || notifications.length === 0) {
        list.innerHTML =
            '<div class="notif-empty">' +
            '<div class="notif-empty-icon">🔔</div>' +
            '<h3>' + notifT("You're all caught up!", "நீங்கள் அனைத்தையும் பார்த்துவிட்டீர்கள்!") + "</h3>" +
            '<p>' + notifT("New farming notifications will appear here.", "புதிய விவசாய அறிவிப்புகள் இங்கே தோன்றும்.") + "</p>" +
            '<a href="/dashboard" class="notif-btn notif-btn-primary">📊 ' + notifT("Go to Dashboard", "டாஷ்போர்டுக்கு செல்ல") + "</a>" +
            "</div>";
        return;
    }
    list.innerHTML = notifications.map(function (n) {
        var icons = {
            scheme: "🏛️", scheme_update: "🔄", price_up: "📈", price_down: "📉",
            irrigation: "💧", fertilizer: "🧪", info: "ℹ️",
        };
        var icon = icons[n.type] || "ℹ️";
        var catBadge = notifCatBadge(n.category);
        var priBadge = notifPriorityBadge(n.priority);
        var timeStr = notifTimeAgo(n.created_at);
        var unreadCls = n.is_read ? "" : " unread";
        var readBtn = n.is_read ? "" :
            '<button class="notif-action-btn" onclick="notifMarkRead(\'' + n.id + '\')">✅ ' + notifT("Read", "படித்தது") + "</button>";
        return '<div class="notif-card' + unreadCls + '">' +
            '<div class="notif-card-icon">' + icon + "</div>" +
            '<div class="notif-card-body">' +
            '<div class="notif-card-title">' + notifEscape(n.title) + "</div>" +
            '<div class="notif-card-msg">' + notifEscape(n.message) + "</div>" +
            '<div class="notif-card-footer">' +
            catBadge + priBadge +
            '<span class="notif-time">' + timeStr + "</span>" +
            "</div></div>" +
            '<div class="notif-card-actions">' +
            readBtn +
            '<button class="notif-action-btn danger" onclick="notifClear(\'' + n.id + '\')">✕</button>' +
            "</div></div>";
    }).join("");
}

function notifCatBadge(cat) {
    var labels = {
        scheme: notifT("Scheme", "திட்டம்"),
        market: notifT("Market", "சந்தை"),
        irrigation: notifT("Irrigation", "நீர்ப்பாசனம்"),
        fertilizer: notifT("Fertilizer", "உரம்"),
    };
    var cls = "notif-badge-" + (cat || "info");
    return '<span class="notif-badge ' + cls + '">' + (labels[cat] || notifT("Info", "தகவல்")) + "</span>";
}

function notifPriorityBadge(pri) {
    var labels = {
        high: notifT("High", "அதிக"),
        medium: notifT("Medium", "நடுத்தர"),
        low: notifT("Low", "குறைந்த"),
    };
    return '<span class="notif-priority notif-priority-' + pri + '">' + (labels[pri] || pri) + "</span>";
}

function notifTimeAgo(dateStr) {
    if (!dateStr) return "";
    var now = new Date();
    var d = new Date(dateStr);
    if (isNaN(d.getTime())) return dateStr;
    var diff = Math.floor((now - d) / 1000);
    if (diff < 60) return notifT("Just now", "இப்போது");
    if (diff < 3600) return Math.floor(diff / 60) + "m";
    if (diff < 86400) return Math.floor(diff / 3600) + "h";
    if (diff < 172800) return notifT("Yesterday", "நேற்று");
    if (diff < 604800) return Math.floor(diff / 86400) + "d";
    return d.toLocaleDateString();
}

function notifEscape(s) {
    if (typeof s !== "string") return s || "";
    var d = document.createElement("div");
    d.textContent = s;
    return d.innerHTML;
}

/* Mark single as read */
function notifMarkRead(id) {
    fetch("/api/notifications/read", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ notification_id: id }),
    })
        .then(function (r) { return r.json(); })
        .then(function (res) {
            if (res.success) notifLoadList();
        });
}

/* Mark all read */
function notifMarkAllRead() {
    fetch("/api/notifications/read", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({}),
    })
        .then(function (r) { return r.json(); })
        .then(function (res) {
            if (res.success) {
                notifLoadList();
                notifToast(notifT("All marked as read", "அனைத்தும் படித்ததாக குறிக்கப்பட்டது"), "success");
            }
        });
}

/* Clear single */
function notifClear(id) {
    fetch("/api/notifications/clear", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ notification_id: id }),
    })
        .then(function (r) { return r.json(); })
        .then(function (res) {
            if (res.success) notifLoadList();
        });
}

/* Clear all */
function notifClearAll() {
    if (!confirm(notifT("Clear all notifications?", "அனைத்து அறிவிப்புகளையும் அழிக்கவா?"))) return;
    fetch("/api/notifications/clear", { method: "POST" })
        .then(function (r) { return r.json(); })
        .then(function (res) {
            if (res.success) {
                notifLoadList();
                notifToast(notifT("All notifications cleared", "அனைத்து அறிவிப்புகளும் அழிக்கப்பட்டன"), "success");
            }
        });
}

File: notifications.py
from flask import Blueprint, render_template, request, jsonify, session
from utils.auth import login_required
from models.notification import Notification
from services.notification_service import NotificationService
import traceback
from datetime import datetime

notifications_bp = Blueprint("notifications", __name__)


@notifications_bp.route("/notifications")
@login_required
def index():
    return render_template("notifications.html", lang=session.get("lang", "en"))


@notifications_bp.route("/api/notifications")
@login_required
def get_notifications():
    try:
        user_id = session.get("user_id")
        page = int(request.args.get("page", 1))
        per_page = int(request.args.get("per_page", 50))
        category = request.args.get("category", "")
        all_param = request.args.get("all", "0")

        query = {"user_id": user_id}
        if category:
            query["category"] = category

        total = Notification.get_collection().count_documents(query)
        notifs = list(Notification.get_collection().find(query).sort("created_at", -1).skip((page - 1) * per_page).limit(per_page))
        unread = Notification.count_unread(user_id)

        return jsonify({
            "success": True,
            "notifications": [Notification(n).to_dict() for n in notifs],
            "unread_count": unread,
            "total": total,
            "page": page,
        })
    except Exception as e:
        print(f"[Notifications Error] list: {traceback.format_exc()}")
        return jsonify({"success": False, "error": "Failed to load notifications"}), 500


@notifications_bp.route("/api/notifications/generate", methods=["POST"])
@login_required
def generate():
    try:
        user_id = session.get("user_id")
        lang = session.get("lang", "en")
        NotificationService.generate_all(user_id, lang)
        return jsonify({"success": True, "message": "Notifications generated"})
    except Exception as e:
        print(f"[Notifications Error] generate: {traceback.format_exc()}")
        return jsonify({"success": False, "error": "Failed to generate"}), 500


@notifications_bp.route("/api/notifications/read", methods=["POST"])
@login_required
def mark_read():
    try:
        data = request.get_json(silent=True) or {}
        notif_id = data.get("notification_id")
        if notif_id:
            notif = Notification.find_by_id(notif_id)
            if notif:
                notif.mark_read()
        else:
            user_id = session.get("user_id")
            Notification.mark_all_read(user_id)
        return jsonify({"success": True})
    except Exception as e:
        return jsonify({"success": False, "error": "Failed"}), 500


@notifications_bp.route("/api/notifications/clear", methods=["POST"])
@login_required
def clear():
    try:
        data = request.get_json(silent=True) or {}
        user_id = session.get("user_id")
        notif_id = data.get("notification_id")
        if notif_id:
            notif = Notification.find_by_id(notif_id)
            if notif and notif.user_id == user_id:
                Notification.get_collection().delete_one({"_id": notif.id})
        else:
            Notification.clear_all(user_id)
        return jsonify({"success": True})
    except Exception as e:
        return jsonify({"success": False, "error": "Failed"}), 500


@notifications_bp.route("/api/notifications/unread")
@login_required
def unread_count():
    try:
        user_id = session.get("user_id")
        count = Notification.count_unread(user_id)
        return jsonify({"success": True, "count": count})
    except Exception as e:
        return jsonify({"success": False, "error": "Failed"}), 500
